packages/dart/lib/src/telemetry_processing/telemetry_buffer.dart

```diff
@@ -0,0 +1,111 @@
+import 'dart:async';
+
+import '../../sentry.dart';
+import 'envelope_builder.dart';
+import 'telemetry_item.dart';
+
+abstract class TelemetryBuffer<T extends TelemetryItem> {
+  void add(T item);
+  FutureOr<void> flush();
+}
+
+/// Holds both raw item and encoded bytes for size tracking and grouping.
+class EncodedTelemetryItem<T extends TelemetryItem> {
+  final T item;
+  final List<int> encoded;
+
+  EncodedTelemetryItem(this.item, this.encoded);
+}
+
+/// In-memory buffer with time and size-based flushing.
+class InMemoryTelemetryBuffer<T extends TelemetryItem>
```

Comment:
Cocoa has additionaly seperated the InMemory Storage from the logic when to dispatch the buffered elements. Not sure this is needed here, here's the [PR](https://github.com/getsentry/sentry-cocoa/pull/7017)